marconi-chain-index/bench/BenchQueries.hs

```diff
@@ -0,0 +1,205 @@
+import System.FilePath ((</>))
+import Test.Tasty.Bench (bench, bgroup, defaultMain, nfIO)
+import Text.RawString.QQ (r)
+
+data FrequencyRow a = FrequencyRow
+    { _frequencyRowValue :: !a
+    , _frequencyRowFreq  :: !Int
+    } deriving (Generic)
+
+instance (FromField a) => FromRow (FrequencyRow a) where
+    fromRow = FrequencyRow <$> field <*> field
+
+instance (ToField a) => ToRow (FrequencyRow a) where
+  toRow (FrequencyRow ad f) = toRow (ad, f)
+
+main :: IO ()
+main = do
+    -- TODO Need a better way to pass those params in the benchmark.
+    -- Look into "Custom command-line options" in the tasty-bench README.
+    nodeSocketPath <- getEnv "CARDANO_NODE_SOCKET_PATH"
+    databaseDir <- getEnv "MARCONI_DB_DIRECTORY_PATH"
+
+    indexerTVar <- atomically (newEmptyTMVar :: STM (TMVar UtxoIndexer) )
+
+    -- Run concurrently the indexing and the testing. The testing is only run once the indexing is
+    -- fully synced with the local running node.
+    race_ (runIndexerSyncing databaseDir nodeSocketPath indexerTVar) $ do
+        putStrLn "Waiting for indexer to be fully synced..."
+        waitUntilSynced databaseDir nodeSocketPath
+        putStrLn "Finished syncing!"
+        tests databaseDir indexerTVar
+
+-- | Run the IO code which syncs the Marconi Utxo indexer with the local running Cardano node.
+runIndexerSyncing
+    :: FilePath
+    -- ^ Marconi indexer database directory
+    -> FilePath
+    -- ^ Local node socket file path
+    -> TMVar UtxoIndexer
+    -> IO ()
+runIndexerSyncing databaseDir nodeSocketPath indexerTVar = do
+    let callbackUtxoIndexer :: UtxoIndexer -> IO ()
+        callbackUtxoIndexer utxoIndexer = atomically $ writeTMVar indexerTVar utxoIndexer
+    (chainPointsToResumeFrom, coordinator) <-
+      startIndexers
+          [ ( utxoWorker callbackUtxoIndexer Nothing
+            -- TODO This hardcoded value needs to be the same value passed through the CLI params in
+            -- the `marconi-chain-index` exe.
+            , databaseDir </> "utxo.db"
+            )
+          ]
+    let indexers = mkIndexerStream coordinator
+    withChainSyncEventStream
+        nodeSocketPath
+        (C.Testnet $ C.NetworkMagic 1) -- TODO Needs to be passed a CLI param
+        chainPointsToResumeFrom
+        indexers
+
+tests
+    :: FilePath
+    -- ^ Marconi indexers database directory path
+    -> TMVar UtxoIndexer
+    -> IO ()
+tests databaseDir indexerTVar = do
+    utxoIndexer <- atomically $ readTMVar indexerTVar
+
+    -- TODO This hardcoded value needs to be the same value that was passed through the CLI params
+    -- in the `marconi-chain-index` exe.
+    c <- SQL.open $ databaseDir </> "utxo.db"
+
+    (addressesWithMostUtxos :: [FrequencyRow C.AddressAny]) <- SQL.query c
+        [r|SELECT address, COUNT(address) as frequency
+           FROM unspent_transactions u
+           LEFT JOIN spent s
+           ON u.txId = s.txInTxId
+             AND u.txId IS NOT NULL
```

Comment:
The `is not null` is not required because null compared to anything (even null itself) evaluates to null, so `u.txId = s.txInTxId` well be null even when both are null. Also, the txId is already set defined `not null` in the sql table's definition.

Reply:
Good to know :) Thanks!